### PR TITLE
Migrate build system from Maven to Gradle

### PR DESCRIPTION
**Pull request by cliu123**
##  opendistro-for-elasticsearch/security pull request intake form
_Please provide as much details as possible to get feedback/acceptance on your PR quickly_

 
 1. __Category:__ (_Enhancement, New feature, Bug fix, Test fix, Refactoring, Maintenance, Documentation_)
 Maintenance
 
 
 2. __Github Issue # or road-map entry, if available:__

 3. __Description of changes:__

 4. __Why these changes are required?__ 
Security plugin has 2 build systems: Maven and Gradle(building `.rpm` and `.deb` packages). It's necessary to combine them into a single one(Gradle).

 5. __What is the old behavior before changes and new behavior after changes?__ (_Please add any example/logs/screen-shot if available_)

 6. __Testing done:__ (_Please provide details of testing done: Unit testing, integration testing and manual testing_)

- Unit Tests
- Manual Testing:
    - Ran `./gradlew clean build --no-daemon -Dbuild.snapshot=false -x test`, artifacts are generated in `gradle-build/distributions/`
    - The following artifacts are generated:
         - `opendistro_security-1.13.1.0.jar`
         - `opendistro-security-1.13.1.0.zip`
         - `opendistro-security-1.13.1.0-securityadmin-standalone.zip`
         - `opendistro-security-1.13.1.0-securityadmin-standalone.tar.gz` 
    - Installed security plugin generated by Gradle on ES. Securify configurations are loaded into `elasticsearch-7.10.2/plugins/opendistro_security/securityconfig`
    - Ran `install_demo_configuration.sh`, security settings are added into `elasticsearch.yml` file.
    - The plugin folder name in `elasticsearch-7.10.2/plugins` is `opendistro_security`
 
 
 7. __TO-DOs, if any:__ (_Please describe pending items and provide Github issues# for each of them_)

 8. __Is it backport from main branch?__ (_If yes, please add backport PR # and commits #_)

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -16,10 +16,10 @@ jobs:
 
     steps:
 
-    - name: Set up JDK 11
+    - name: Set up JDK 14
```
Comment (cliu123) [inline .github/workflows/ci.yml:19]: Gradle > 6.3 supports Java 14

**Review by vrozov — CHANGES_REQUESTED** [review]
Please limit scope of the PR just to migration from maven to gradle build system. Anything not directly related to maven->gradle migration such as UT, java doc, JDK version upgrade should be part of a separate PR(s).

File: gradle/wrapper/gradle-wrapper.properties [hunk 2]
```diff
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-4.10.2-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.8-bin.zip
```
Comment (vrozov) [inline gradle/wrapper/gradle-wrapper.properties:3]: If gradle version upgrade is necessary, please move that to a separate PR as well.

File: build.gradle [hunk 3]
```diff
@@ -1,8 +1,270 @@
...
+
+configurations.testCompile {
+    if (it.state != Configuration.State.UNRESOLVED) return
+    resolutionStrategy {
+        force 'org.apache.kafka:kafka-clients:2.0.1:test'
+    }
+}
+
+dependencies {
+    compile('org.elasticsearch.plugin:transport-netty4-client:7.10.2') {
```
Comment (sujithvm) [inline build.gradle:120]: `compile` is deprecated in favor of `implementation`. Can we use that instead?

File: build.gradle [hunk 4]
```diff
@@ -1,8 +1,270 @@
...
+        exclude group: 'org.apache.logging.log4j', module: 'log4j-api'
+    }
+    compile('org.ldaptive:ldaptive:1.2.3') {
+        exclude group: 'commons-cli', module: 'commons-cli'
+        exclude group: 'org.slf4j', module: 'slf4j-api'
+    }
+    compile('org.apache.httpcomponents:httpclient-cache:4.5.3') {
+        exclude group: 'org.apache.httpcomponents', module: 'httpcore'
+    }
+    compile('org.elasticsearch.client:elasticsearch-rest-high-level-client:7.10.2') {
```
Comment (sujithvm) [inline build.gradle:143]: Please define variable and reuse for same versions. Eg: 7.10.2 is used for a couple of dependencies.

File: build.gradle [hunk 5]
```diff
@@ -1,8 +1,270 @@
...
+    testImplementation 'junit:junit:4.12'
+    testImplementation 'org.apache.httpcomponents:fluent-hc:4.5.3'
+    testImplementation('org.elasticsearch.plugin:reindex-client:7.10.2') {
+        exclude group: 'org.elasticsearch', module: 'elasticsearch-ssl-config'
+        exclude group: 'org.apache.httpcomponents', module: 'httpclient'
+    }
+    testImplementation 'org.elasticsearch:elasticsearch-ssl-config:7.10.2'
+    testImplementation 'org.elasticsearch.plugin:percolator-client:7.10.2'
+    compile 'org.elasticsearch.plugin:lang-mustache-client:7.10.2'
+    compile 'org.elasticsearch.plugin:parent-join-client:7.10.2'
```
Comment (sujithvm) [inline build.gradle:203]: Group all testImplementation, implementation, compileOnly, runtimeOnly dependencies

File: build.gradle [hunk 6]
```diff
@@ -1,8 +1,270 @@
...
+ext {
+    projectSubstitutions = [:]
+    licenseFile = rootProject.file('LICENSE.txt')
+    noticeFile = rootProject.file('NOTICE.txt')
+}
+
+configurations.all {
+    if (it.state != Configuration.State.UNRESOLVED) return
+    resolutionStrategy {
+        force 'commons-codec:commons-codec:1.14'
```
Comment (sujithvm) [inline build.gradle:85]: Do all the dependencies need force?

File: gradle/wrapper/gradle-wrapper.properties (hunk 2 above)
Comment (cliu123) [reply gradle/wrapper/gradle-wrapper.properties:3]: > Please limit scope of the PR just to migration from maven to gradle build system. Anything not directly related to maven->gradle migration such as UT, java doc, JDK version upgrade should be part of a separate PR(s).

I've excluded java doc from compilatoin, I can remove the change in java doc. 
But all other changes are necessary. If we move all these changes to separate PRs, we'd have to create quite a few PRs and merge them before this PR. 
Would that help if I modify the title of the PR and add some description for those changes in the PR description to clarify the scope of the PR?

**Comment by cliu123** [comment]
> Please limit scope of the PR just to migration from maven to gradle build system. Anything not directly related to maven->gradle migration such as UT, java doc, JDK version upgrade should be part of a separate PR(s).

I've excluded java doc from compilatoin, I can remove the change in java doc.
But all other changes are necessary. If we move all these changes to separate PRs, we'd have to create quite a few PRs and merge them before this PR.
Would that help if I modify the title of the PR and add some description for those changes in the PR description to clarify the scope of the PR?

**Comment by vrozov** [comment]
Change PR and commit title to something like "migrate build system from maven to gradle"

**Comment by vrozov** [comment]
Changing PR title is necessary to avoid confusion regarding the goal of the PR, but it will not help with other issues. If version of gradle needs to be upgraded, it should be a separate PR and commit as in one large PR it is impossible to distinguish changes required for gradle upgrade and changes required for migrating from maven to gradle. The same applies to all other unrelated to maven -> gradle changes.

File: build.gradle (hunk 3 above)
Comment (cliu123) [reply build.gradle:120]: Replacing all `compile` with `implementation` fails `compileJava` task. Some of them indeed can be replaced. I didn't change them for consistency.

File: build.gradle (hunk 6 above)
Comment (cliu123) [reply build.gradle:85]: Yes, all these need force because otherwise some dependency conflicts come out.

File: build.gradle (hunk 4 above)
Comment (cliu123) [reply build.gradle]: Done.

File: build.gradle (hunk 5 above)
Comment (cliu123) [reply build.gradle]: By group, do you mean put dependencies with same scope together?

Comment (cliu123) [reply build.gradle]: Done

**Comment by cliu123** [comment]
Will re-open later if needed.
